render-ws-java-client/src/main/java/org/janelia/render/client/newsolver/solvers/affine/MultiSemPreAligner.java

```diff
@@ -0,0 +1,184 @@
+package org.janelia.render.client.newsolver.solvers.affine;
+
+import mpicbg.models.Affine2D;
+import mpicbg.models.ErrorStatistic;
+import mpicbg.models.Model;
+import mpicbg.models.PointMatch;
+import mpicbg.models.Tile;
+import mpicbg.models.TileConfiguration;
+import mpicbg.models.TileUtil;
+import net.imglib2.util.Pair;
+import net.imglib2.util.ValuePair;
+import org.janelia.alignment.match.CanvasMatchResult;
+import org.janelia.alignment.match.CanvasMatches;
+import org.janelia.alignment.spec.ResolvedTileSpecCollection;
+import org.janelia.alignment.spec.TileSpec;
+import org.janelia.render.client.solver.SolveTools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.DoubleSummaryStatistics;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Coarsely aligns a multi-SEM stack by treating each mFOV layer as a single tile.
+ *
+ * @param <M> the model type
+ *
+ * @author Michael Innerberger
+ */
+public class MultiSemPreAligner<M extends Model<M> & Affine2D<M>> implements Serializable {
+
+	// Note: this assumes a specific tileId format:
+	// <cut>_<mFov>_<sFov>_<day>_<time>.<z>.0
+	private static final Pattern TILE_ID_SEPARATOR = Pattern.compile("[_.]");
+
+	private final M model;
+	private final double maxAllowedError;
+	private final int maxIterations;
+	private final int maxPlateauWidth;
+	private final int numThreads;
+	private final int maxNumMatches;
+
+
+	public MultiSemPreAligner(
+			final M model,
+			final double maxAllowedError,
+			final int maxIterations,
+			final int maxPlateauWidth,
+			final int numThreads,
+			final int maxNumMatches
+	) {
+		this.model = model;
+		this.maxAllowedError = maxAllowedError;
+		this.maxIterations = maxIterations;
+		this.maxPlateauWidth = maxPlateauWidth;
+		this.numThreads = numThreads;
+		this.maxNumMatches = maxNumMatches;
+	}
+
+	public Map<String, M> preAlign(
+			final ResolvedTileSpecCollection rtsc,
+			final List<CanvasMatches> canvasMatches
+	) throws IOException, ExecutionException, InterruptedException {
+
+		// initialize and connect models for each mFOV layer
+		final Map<String, String> tileIdToMFovLayer = rtsc.getTileSpecs().stream()
+				.map(TileSpec::getTileId)
+				.collect(Collectors.toMap(
+						tileId -> tileId,
+						MultiSemPreAligner::extractMFovLayerId
+				));
+		final Map<String, Tile<M>> mFovLayerToTile = initializeAndConnectTiles(rtsc, tileIdToMFovLayer, canvasMatches);
```

Comment:
Maybe we can do this while retrieving just the data for the first 2 z layers and then looping through grabbing data for one more layer at time while building/connecting the MFOV tiles.

Reply:
I'm just realizing that all your comments suggest that interleaving retrieval from the database with the algorithm, whereas I tried to keep database communication and algorithmic details separate. Let's talk about how to tackle this problem on a general scale.